src/Domain/Operation/isOperationPermitted.ts

```diff
@@ -0,0 +1,22 @@
+import { UserPermissions } from '../User/UserWithPermissions'
+import { Operation } from './Operation'
+import { Resource } from '../Resource/Resource'
+
+// todo: implement based on what auth is capable of returning; + tests
+/**
+ * Determines if the given permissions allow executing the requested operation.
+ */
+export function isOperationPermitted({
```

Comment:
same as above - please turn all code into classes instead of functions. IMHO functional programming does not go well with DI, OOP and Hexagonal Architecture like we do have in our server side projects

Reply:
This response relates to these comments:

https://github.com/standardnotes/files/pull/3#discussion_r594916102
https://github.com/standardnotes/files/pull/3#discussion_r594916271
https://github.com/standardnotes/files/pull/3#discussion_r594916446
https://github.com/standardnotes/files/pull/3#discussion_r594908953

***

I am not proposing introducing FP or any revolutionary changes, but I must present an argument against this.

My view is: it's best to avoid premature abstraction, start as simple as possible and build on/complicate that if necessary. This applies in any paradigm.

The simplest unit of abstraction in most languages (OOP, functional, mixed) is a function. The simplest functions are pure, stateless. They go well with everything. Because they are stateless and have clearly defined inputs and outputs, they are as testable as it gets. 

I think that if a language provides means to use simple functions without forcing an additional layer of abstraction up-front, we should use that.

Most DI frameworks like inversify are indeed designed primarily around OOP and only [provide a crude support for functions](https://github.com/inversify/InversifyJS/blob/master/wiki/recipes.md#injecting-dependencies-into-a-function)* but that's a bit beside the point.

If we need to play nice with the DI, we can always add the abstractions required by it. Subtracting though, in my experience, is very rarely (or never) done, leading to a proliferation of one-off abstractions which add no practical value.

When I see these two possibilities:

1. Creating a pure and stateless function `do` and calling it where I need it.

2. Doing 1., but also wrapping the function in a stateful `Doer` class which implements a `DoerInterface`, creating and passing an object around with DI, and only then calling the wrapped function.

I default to the first option, especially if there is no sign that more `DoerInterface` implementers will appear in the future.

And if a need arises for one, I can always add the abstractions then and there.

\2. feels like forcing the DI to do unnecessary work. There is no return on the introduced complexity.

\* BTW I have no experience with TypeScript DI frameworks, but inversify itself so far feels rather crude, even compared to [Guice](https://github.com/google/guice) -- this is quite surprising, given the difference in flexibility between Java and TypeScript.

***

In general, in FP vs OOP or X vs Y discussions I am on the side of simplicity and pragmatism, as I perceive them. I am always open to discussion and sensible arguments to change my perceptions if they turn out to be misguided.